### PR TITLE
B#34 emptymapping

### LINKED ISSUE #34: hand_finder cannot work with empty mapping
The current hand finder is designed to store mapping (naming) of the hands in a dictionary with serial_id as key, and joint_prefix uses mapping as key. 
This is problematic if the mapping is empty.

Also, there are some mix-up between mapping and joint_prefix, sometimes joint_prefix is generated from mapping + "_"

Finally config folder name should be separate from mapping.

I created a branch in my fork (https://github.com/ubi-agni/sr_core/tree/F_handfinder_namespace) , that adds a config_dir, relies on serial_id as the key (as this is never empty and always unique), handles empty prefix, fixes the mix-up and adds tests.

If you like the suggested solution, I can adapt part of the other software (sr_edc_launch and friends) to work with the hand_finder. Theoretically the other software only using prefix and mapping and no config_dir, and should not need modification when calling the hand finder.

### PR DESCRIPTION
**Pull request by guihomework**
fixes #34 

First commits add tests that highlight issues
Next commits solve one bug at a time (almost)
- bug if no hand mapping at all and no robot description
- bug mix-up of hand_id and joint_prefix
- bug with wrong keys used
- empty hand_id support
- fixes after tested on real hand (robot_description)
- bug with hand param not looked globally
- bug with calibrate lockup at start (race condition)

### CODE REVIEW

File: sr_utilities/test/test_hand_finder.cpp [hunk 1]
```diff
@@ -54,6 +121,134 @@ TEST(SrHandFinder, hand_present_test)
...
+{
+  ros::NodeHandle nh;
+  nh.setParam("hand/mapping/1", "");
+  nh.setParam("hand/joint_prefix/1", "");
+  const string joint_names[] = {"FFJ1", "FFJ2", "FFJ3", "FFJ4", "MFJ1", "MFJ2", "MFJ3", "MFJ4",
+                                "RFJ1", "RFJ2", "RFJ3", "RFJ4", "LFJ1", "LFJ2", "LFJ3", "LFJ4", "LFJ5",
+                                "THJ1", "THJ2", "THJ3", "THJ4", "THJ5", "WRJ1", "WRJ2"};
+  shadow_robot::SrHandFinder hand_finder;
+  map<string, vector<string> > hand_joints(hand_finder.get_joints());
+  for (map<string, vector<string> >::const_iterator iter = hand_joints.begin(); iter != hand_joints.end(); ++iter)
```
Comment (toliver) [inline sr_utilities/test/test_hand_finder.cpp]: I think that in this case `hand_joints` will be an empty map, isn't it? So maybe we should check:

``` cpp
 ASSERT_TRUE(hand_joints.empy());
```

File: sr_utilities/test/test_hand_finder.cpp [hunk 2]
```diff
@@ -54,6 +121,134 @@ TEST(SrHandFinder, hand_present_test)
...
+{
+  ros::NodeHandle nh;
+  nh.setParam("hand/mapping/1", "rh");
+  nh.setParam("hand/joint_prefix/1", "");
+  const string joint_names[] = {"FFJ1", "FFJ2", "FFJ3", "FFJ4", "MFJ1", "MFJ2", "MFJ3", "MFJ4",
+                                "RFJ1", "RFJ2", "RFJ3", "RFJ4", "LFJ1", "LFJ2", "LFJ3", "LFJ4", "LFJ5",
+                                "THJ1", "THJ2", "THJ3", "THJ4", "THJ5", "WRJ1", "WRJ2"};
+  shadow_robot::SrHandFinder hand_finder;
+  map<string, vector<string> > hand_joints(hand_finder.get_joints());
+  for (map<string, vector<string> >::const_iterator iter = hand_joints.begin(); iter != hand_joints.end(); ++iter)
```
Comment (toliver) [inline sr_utilities/test/test_hand_finder.cpp]: Same here.

File: sr_utilities/test/test_hand_finder.py [hunk 3]
```diff
@@ -33,33 +49,203 @@ def test_no_hand_finder(self):
...
+
+        ctrl_tun_host_control_paths = hand_finder.get_hand_control_tuning().host_control['right']
+
+        self.assertEqual(len(ctrl_tun_host_control_paths), len(controller_params),
+                         "incorrect number of host controllers param")
+        for controller_path, controller_param in zip(ctrl_tun_host_control_paths, controller_params):
+            self.assertEqual(controller_path, self.ethercat_path + "/controls/host/right/" + controller_param,
+                             "incorrect controller config file")
+
+    def test_one_hand_finder_no_ns(self):
```
Comment (toliver) [inline sr_utilities/test/test_hand_finder.py]: Maybe we should call it no_mapping for consistency.

File: sr_utilities/test/test_hand_finder.cpp (hunk 1 above)
Comment (guihomework) [inline sr_utilities/test/test_hand_finder.cpp]: why would the hand_joints be empty ? the joint_prefix exists (it is set to an empty string but exists), so there will be joint_names generated without any joint_prefix at the front, and the tests, verifies that.

File: sr_utilities/test/test_hand_finder.cpp (hunk 2 above)
Comment (guihomework) [inline sr_utilities/test/test_hand_finder.cpp]: same remark, an empty prefix is not an empty parameter.

File: sr_utilities/test/test_hand_finder.py (hunk 3 above)
Comment (guihomework) [inline sr_utilities/test/test_hand_finder.py]: indeed, no_mapping is better for consistency,can you fix it while merging ?

File: sr_utilities/test/test_hand_finder.cpp (hunk 1 above)
Comment (toliver) [inline sr_utilities/test/test_hand_finder.cpp]: You are right indeed. ;)

**Comment by toliver** [comment]
Very nice set of tests. I'm going to merge after fixing the conflicts.

**Comment by toliver** [comment]
The conflicts are not trivial to merge, so we will leave it pending until we have a slot for this task.

**Comment by guihomework** [comment]
I feared that after 5 months. Do you think I have a chance to solve that on my own ?

**Comment by toliver** [comment]
Yes, the conflicts are derived from lint error fixes mainly, so it shouldn't be very difficult to fix. And with the tests it should be safe to check everything works.

**Comment by guihomework** [comment]
Working on it, but it is far from trivial, you added some asserts that I did not have and I added some tests that would require to have those asserts as well. My mergetool gets mad and does not help much. Please be patient for the new rebased PR.

**Comment by ugocupcic** [comment]
@guihomework "shadow-jenkins" is a bot that spawns this message from our build server! Sorry about that, he never shuts up :P

**Comment by guihomework** [comment]
I also wanted to announce that I started working on the topic, so that we do not work twice. 

The rebase is rather complex, because I have also to understand what you expect with some of the tests, especially when there is no robot_description or when there is one. This is new and rebasing and testing at the same time takes more time.

It is definitely not a lint problem

**Comment by guihomework** [comment]
Well, the rebase was hard but I made it, introducing some lint errors at the same time due to strange styling of namespace in cpp and so on, but fixed by some new commits. 

All test pass now, and tests are more complete. 

Hopefully this will be merged soon otherwise the same rebasing issues will occur.

**Pull request merged**